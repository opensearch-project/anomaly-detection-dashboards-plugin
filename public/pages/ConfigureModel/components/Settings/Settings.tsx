/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  EuiCompressedFieldNumber,
  EuiFlexGroup,
  EuiFlexItem,
  EuiText,
} from '@elastic/eui';
import { Field, FieldProps, useFormikContext } from 'formik';
import React, { Fragment } from 'react';
import {
  isInvalid,
  getError,
  validatePositiveInteger,
  validateMultipleOf,
  validateEmptyOrPositiveInteger,
  validateEmptyOrNonNegativeInteger,
} from '../../../../utils/utils';
import { AD_DOCS_LINK } from '../../../../utils/constants';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';

export const Settings = () => {
  const formik = useFormikContext();
  const numberFieldWidth = 140;

  // Custom validation function that has access to form values
  const validateFrequency = (value: any) => {
    return validateMultipleOf(value, formik.values.interval);
  };

  return (
    <Fragment>
      <Field name="interval" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps) => {
          // Keep frequency in sync when the interval changes (only when they started equal).
          // This reduces UX friction for users who don't care to tweak frequency while editing interval.
          const handleIntervalChange = (
            event: React.ChangeEvent<HTMLInputElement>
          ) => {
            const raw = event.target.value;
            const nextInterval = raw === '' ? '' : Number(raw);

            // remember the values before we mutate anything
            const prevInterval = field.value;
            const prevFrequency = form.values.frequency;

            // push the new interval into Formik
            form.setFieldValue(field.name, nextInterval);

            // if frequency was just mirroring interval, keep it in lockstep
            const prevIntervalNumber = Number(prevInterval);
            const prevFrequencyNumber = Number(prevFrequency);
            const wasSynced =
              !Number.isNaN(prevIntervalNumber) &&
              prevIntervalNumber === prevFrequencyNumber;

            if (wasSynced) {
              form.setFieldValue('frequency', nextInterval);
            }
          };

          return (
            <EuiFlexGroup>
              <EuiFlexItem style={{ maxWidth: '70%' }}>
                <FormattedFormRow
                  fullWidth
                  title="Interval"
                  hint={[
                    `Interval sets the time window for summarizing and modeling data (e.g., 5 min to 1 hr), where too small creates noise, higher cost, and overreaction to fluctuations, while too large smooths out anomalies and delays detection.`,
                  ]}
                  hintLink={`${AD_DOCS_LINK}`}
                  isInvalid={isInvalid(field.name, form)}
                  error={getError(field.name, form)}
                >
                  <EuiFlexGroup gutterSize="s" alignItems="center">
                    <EuiFlexItem grow={false}>
                      <EuiCompressedFieldNumber
                        name="detectionInterval"
                        id="detectionInterval"
                        placeholder="Interval"
                        data-test-subj="detectionInterval"
                        min={1}
                        style={{ width: numberFieldWidth }}
                        {...field}
                        onChange={handleIntervalChange}
                      />
                    </EuiFlexItem>
                    <EuiFlexItem>
                      <EuiText>
                        <p className="minutes">minutes</p>
                      </EuiText>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                </FormattedFormRow>
              </EuiFlexItem>
            </EuiFlexGroup>
          );
        }}
      </Field>

      <Field name="frequency" validate={validateFrequency}>
        {({ field, form }: FieldProps) => (
          <EuiFlexGroup style={{ marginTop: '16px' }}>
            <EuiFlexItem style={{ maxWidth: '70%' }}>
              <FormattedFormRow
                fullWidth
                title="Frequency"
                hint={[
                  `Frequency sets how often the detector queries and scores data—i.e., how often alerts may fire. It must be a multiple of the interval and defaults to the same value. Choosing a longer frequency than the interval can improve efficiency by batching short intervals when ultra-fast alerting isn’t required (e.g., monitoring daily counts of anomalies across detectors) and is also useful for infrequent or batch log ingestion where data arrives irregularly or in bulk (e.g., a single daily ingestion spanning the entire day).`,
                ]}
                hintLink={`${AD_DOCS_LINK}`}
                isInvalid={isInvalid(field.name, form)}
                error={getError(field.name, form)}
              >
                <EuiFlexGroup gutterSize="s" alignItems="center">
                  <EuiFlexItem grow={false}>
                    <EuiCompressedFieldNumber
                      name="frequency"
                      id="frequency"
                      placeholder="Frequency"
                      data-test-subj="frequency"
                      min={1}
                      style={{ width: numberFieldWidth }}
                      {...field}
                    />
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <EuiText>
                      <p className="minutes">minutes</p>
                    </EuiText>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </FormattedFormRow>
            </EuiFlexItem>
          </EuiFlexGroup>
        )}
      </Field>

      <Field name="windowDelay" validate={validateEmptyOrNonNegativeInteger}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            fullWidth
            title="Window delay"
            hint="Specify a window of delay for a detector to fetch data, if you need to account for extra processing time."
            hintLink={`${AD_DOCS_LINK}`}
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
            style={{ marginTop: '16px' }}
          >
            <EuiFlexGroup gutterSize="s" alignItems="center">
              <EuiFlexItem grow={false}>
                <EuiCompressedFieldNumber
                  name="windowDelay"
                  id="windowDelay"
                  placeholder="Window delay"
                  data-test-subj="windowDelay"
                  style={{ width: numberFieldWidth }}
                  {...field}
                />
              </EuiFlexItem>
              <EuiFlexItem>
                <EuiText>
                  <p className="minutes">minutes</p>
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </FormattedFormRow>
        )}
      </Field>

      <Field name="history" validate={validateEmptyOrPositiveInteger}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            fullWidth
            title="History"
            hint="How far back the model looks for training data. This determines the amount of historical data used to train the anomaly detection model. Minimum history is 40 intervals, maximum is 10,000 intervals, and the default is 40."
            hintLink={`${AD_DOCS_LINK}`}
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
            style={{ marginTop: '16px' }}
          >
            <EuiFlexGroup gutterSize="s" alignItems="center">
              <EuiFlexItem grow={false}>
                <EuiCompressedFieldNumber
                  name="history"
                  id="history"
                  placeholder="History"
                  data-test-subj="history"
                  min={40}
                  max={10000}
                  style={{ width: numberFieldWidth }}
                  {...field}
                />
              </EuiFlexItem>
              <EuiFlexItem>
                <EuiText>
                  <p className="minutes">intervals</p>
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </FormattedFormRow>
        )}
      </Field>
    </Fragment>
  );
};
