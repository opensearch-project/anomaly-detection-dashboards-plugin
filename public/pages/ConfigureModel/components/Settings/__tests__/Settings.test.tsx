/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Settings } from '../Settings';

import { Formik } from 'formik';

describe('<Settings /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('shows error for empty interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Must be a positive integer')).toBeNull();
    // Because that validator is passed in the validate prop,
    // Formik calls it every time the interval field changes
    // or loses focus (and again on form submit).
    await fireEvent.focus(getByPlaceholderText('Interval'));
    await fireEvent.blur(getByPlaceholderText('Interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for invalid interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    await userEvent.type(getByPlaceholderText('Interval'), '-1');
    await fireEvent.blur(getByPlaceholderText('Interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for interval of 0 when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    await userEvent.type(getByPlaceholderText('Interval'), '0');
    await fireEvent.blur(getByPlaceholderText('Interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for invalid window delay when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    await userEvent.type(getByPlaceholderText('Window delay'), '-1');
    await fireEvent.blur(getByPlaceholderText('Window delay'));
    expect(findByText('Must be a non-negative integer')).not.toBeNull();
  });

  describe('handleIntervalChange - frequency auto-adjustment', () => {
    test('syncs frequency when interval and frequency start with same value', async () => {
      const initialValues = {
        interval: 5,
        frequency: 5, // synced with interval
        windowDelay: 0,
        history: 40,
      };

      const { getByPlaceholderText } = render(
        <Formik initialValues={initialValues} onSubmit={jest.fn()}>
          {() => <Settings />}
        </Formik>
      );

      const intervalInput = getByPlaceholderText('Interval');
      const frequencyInput = getByPlaceholderText('Frequency');

      // Initially both should be 5
      expect(intervalInput).toHaveValue(5);
      expect(frequencyInput).toHaveValue(5);

      // Change interval from 5 to 10
      await userEvent.clear(intervalInput);
      // set the final value in one change event
      // when using userEvent.type, typing "10" fires two onChange events: first '1', then '10'
      // In the full file run, the assertion expect(frequencyInput).toHaveValue(10) sometimes
      // executes before that async update finishes.
      // In “maintains sync relationship through multiple changes”, we do more user actions (8 → 12 → 3).
      //  Those extra events/blur cycles effectively give time for the async state/effect/validation to
      //  settle before each assertion, so it passes.
      fireEvent.change(intervalInput, { target: { value: 10 } });
      fireEvent.blur(intervalInput);

      // Frequency should automatically update to 10 (maintaining sync)
      expect(frequencyInput).toHaveValue(10);
    });

    test('does not adjust frequency when values are not initially synced', async () => {
      const initialValues = {
        interval: 5,
        frequency: 10, // not synced with interval
        windowDelay: 0,
        history: 40,
      };

      const { getByPlaceholderText } = render(
        <Formik initialValues={initialValues} onSubmit={jest.fn()}>
          {() => <Settings />}
        </Formik>
      );

      const intervalInput = getByPlaceholderText('Interval');
      const frequencyInput = getByPlaceholderText('Frequency');

      // Initially interval is 5, frequency is 10
      expect(intervalInput).toHaveValue(5);
      expect(frequencyInput).toHaveValue(10);

      // Change interval from 5 to 8
      await userEvent.clear(intervalInput);
      await userEvent.type(intervalInput, '8');
      fireEvent.blur(intervalInput);

      // Frequency should remain 10 (not auto-adjusted since they weren't synced)
      expect(frequencyInput).toHaveValue(10);
    });

    test('maintains sync relationship through multiple changes', async () => {
      const initialValues = {
        interval: 5,
        frequency: 5, // synced
        windowDelay: 0,
        history: 40,
      };

      const { getByPlaceholderText } = render(
        <Formik initialValues={initialValues} onSubmit={jest.fn()}>
          {() => <Settings />}
        </Formik>
      );

      const intervalInput = getByPlaceholderText('Interval');
      const frequencyInput = getByPlaceholderText('Frequency');

      // First change: 5 -> 8
      await userEvent.clear(intervalInput);
      await userEvent.type(intervalInput, '8');
      fireEvent.blur(intervalInput);
      expect(frequencyInput).toHaveValue(8);

      // Second change: 8 -> 12 (should maintain sync)
      await userEvent.clear(intervalInput);
      await userEvent.type(intervalInput, '12');
      fireEvent.blur(intervalInput);
      expect(frequencyInput).toHaveValue(12);

      // Third change: 12 -> 3 (should maintain sync)
      await userEvent.clear(intervalInput);
      await userEvent.type(intervalInput, '3');
      fireEvent.blur(intervalInput);
      expect(frequencyInput).toHaveValue(3);
    });

    test('handles edge case where frequency becomes unsynced after manual change', async () => {
      const initialValues = {
        interval: 5,
        frequency: 5, // initially synced
        windowDelay: 0,
        history: 40,
      };

      const { getByPlaceholderText } = render(
        <Formik initialValues={initialValues} onSubmit={jest.fn()}>
          {() => <Settings />}
        </Formik>
      );

      const intervalInput = getByPlaceholderText('Interval');
      const frequencyInput = getByPlaceholderText('Frequency');

      // Manually change frequency to break sync
      await userEvent.clear(frequencyInput);
      await userEvent.type(frequencyInput, '15');
      fireEvent.blur(frequencyInput);

      // Now change interval - should not auto-adjust frequency since they're no longer synced
      await userEvent.clear(intervalInput);
      await userEvent.type(intervalInput, '10');
      fireEvent.blur(intervalInput);

      // Frequency should remain 15 (not auto-adjusted)
      expect(frequencyInput).toHaveValue(15);
    });
  });
});
