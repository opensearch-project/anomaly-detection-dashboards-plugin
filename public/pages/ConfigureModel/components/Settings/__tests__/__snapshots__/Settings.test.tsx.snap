// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> spec renders the component 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
      style="max-width: 70%;"
    >
      <div
        class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
        hint="Interval sets the time window for summarizing and modeling data (e.g., 5 min to 1 hr), where too small creates noise, higher cost, and overreaction to fluctuations, while too large smooths out anomalies and delays detection."
        id="random_html_id-row"
        title="Interval"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Interval
              </p>
              <br />
              <div
                class="euiText euiText--medium sublabel"
                style="max-width: 400px;"
              >
                Interval sets the time window for summarizing and modeling data (e.g., 5 min to 1 hr), where too small creates noise, higher cost, and overreaction to fluctuations, while too large smooths out anomalies and delays detection.
                 
                <a
                  class="euiLink euiLink--primary"
                  href="https://opensearch.org/docs/latest/observing-your-data/ad/index/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                  <svg
                    aria-hidden="true"
                    aria-label="External link"
                    class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                    />
                  </svg>
                  <span
                    class="euiScreenReaderOnly"
                  >
                    (opens in a new tab or window)
                  </span>
                </a>
              </div>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
            id="random_html_id"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber euiFieldNumber--compressed"
                    data-test-subj="detectionInterval"
                    id="detectionInterval"
                    min="1"
                    name="interval"
                    placeholder="Interval"
                    style="width: 140px;"
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiText euiText--medium"
              >
                <p
                  class="minutes"
                >
                  minutes
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="margin-top: 16px;"
  >
    <div
      class="euiFlexItem"
      style="max-width: 70%;"
    >
      <div
        class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
        hint="Frequency sets how often the detector queries and scores data—i.e., how often alerts may fire. It must be a multiple of the interval and defaults to the same value. Choosing a longer frequency than the interval can improve efficiency by batching short intervals when ultra-fast alerting isn’t required (e.g., monitoring daily counts of anomalies across detectors) and is also useful for infrequent or batch log ingestion where data arrives irregularly or in bulk (e.g., a single daily ingestion spanning the entire day)."
        id="random_html_id-row"
        title="Frequency"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Frequency
              </p>
              <br />
              <div
                class="euiText euiText--medium sublabel"
                style="max-width: 400px;"
              >
                Frequency sets how often the detector queries and scores data—i.e., how often alerts may fire. It must be a multiple of the interval and defaults to the same value. Choosing a longer frequency than the interval can improve efficiency by batching short intervals when ultra-fast alerting isn’t required (e.g., monitoring daily counts of anomalies across detectors) and is also useful for infrequent or batch log ingestion where data arrives irregularly or in bulk (e.g., a single daily ingestion spanning the entire day).
                 
                <a
                  class="euiLink euiLink--primary"
                  href="https://opensearch.org/docs/latest/observing-your-data/ad/index/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                  <svg
                    aria-hidden="true"
                    aria-label="External link"
                    class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                    />
                  </svg>
                  <span
                    class="euiScreenReaderOnly"
                  >
                    (opens in a new tab or window)
                  </span>
                </a>
              </div>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
            id="random_html_id"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber euiFieldNumber--compressed"
                    data-test-subj="frequency"
                    id="frequency"
                    min="1"
                    name="frequency"
                    placeholder="Frequency"
                    style="width: 140px;"
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiText euiText--medium"
              >
                <p
                  class="minutes"
                >
                  minutes
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="Specify a window of delay for a detector to fetch data, if you need to account for extra processing time."
    id="random_html_id-row"
    style="margin-top: 16px;"
    title="Window delay"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            Window delay
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            Specify a window of delay for a detector to fetch data, if you need to account for extra processing time.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/ad/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        id="random_html_id"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--compressed"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldNumber euiFieldNumber--compressed"
                data-test-subj="windowDelay"
                id="windowDelay"
                name="windowDelay"
                placeholder="Window delay"
                style="width: 140px;"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiText euiText--medium"
          >
            <p
              class="minutes"
            >
              minutes
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="How far back the model looks for training data. This determines the amount of historical data used to train the anomaly detection model. Minimum history is 40 intervals, maximum is 10,000 intervals, and the default is 40."
    id="random_html_id-row"
    style="margin-top: 16px;"
    title="History"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            History
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            How far back the model looks for training data. This determines the amount of historical data used to train the anomaly detection model. Minimum history is 40 intervals, maximum is 10,000 intervals, and the default is 40.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/ad/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        id="random_html_id"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--compressed"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldNumber euiFieldNumber--compressed"
                data-test-subj="history"
                id="history"
                max="10000"
                min="40"
                name="history"
                placeholder="History"
                style="width: 140px;"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem"
        >
          <div
            class="euiText euiText--medium"
          >
            <p
              class="minutes"
            >
              intervals
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
